### PR TITLE
Add transpose primitive to JAX interop.

### PR DESCRIPTION
**Pull request by DHHZ**
I found the JAX transpose mechanism quite hard to follow, I've tried adding a bunch of comments to explain but please let me know if anything doesn't make sense, or doesn't fit with your understanding of transpose in JAX.

### CODE REVIEW

File: python/dex/interop/jax.py [hunk 1]
```diff
@@ -271,5 +271,127 @@ def tuple_string(prefix):
...
+  # Add `func_atom` as a variable `linearized` in the context.
+  env = api.insert(func_atom.module, api.as_cstr("linearized"), func_atom)
+
+  # Form lists of the indices in `tangents` which correspond to linear inputs
+  # (which we are expected to transpose w.r.t.) and constant inputs (which we
+  # are not).
+  tangent_input_indices = [
+      i for i, t in enumerate(tangents) if _is_linear_input(t)
+  ]
+  tangent_constant_indices = [
```
Comment (apaszke) [inline python/dex/interop/jax.py:327]: Is that ever non-empty? I don't think so? Or when you say "constant inputs" do you mean the zeros that you've instantiated in the JVP? If yes then let's make the comment more precise.

File: python/dex/interop/jax.py [hunk 2]
```diff
@@ -271,5 +271,127 @@ def tuple_string(prefix):
...
+
+  # Form a lambda which partially evaluates `linearized` at the constant primal
+  # and constant tangent values, with the remaining arguments (i.e. the linear
+  # input tangents) combined into a tuple, and then transpose the lambda.
+  #
+  # For a three-input primal function with constant input for the tangent
+  # parameter at index 1, the evaluated string should look like:
+  # ```
+  # \ x0 x1 x2 u1 ct.
+  #   transposeLinear (\(t0, t2). linearized x0 x1 x2 t0 u1 t2) ct
```
Comment (apaszke) [inline python/dex/interop/jax.py:344]: This works, although one downside of this trick is that it effectively makes all Dex primitives do checkpointing as if they were decorated with `jax.remat`.

**Review by apaszke — COMMENTED** [review]
Looks very good! I only have one minor question and a potential challenge to efficiency, but this is pretty much ready to merge.

File: python/dex/interop/jax.py (hunk 1 above)
Comment (DHHZ) [reply python/dex/interop/jax.py:328]: It's the latter, added a further comment to clarify.

The linear / constant input terminology I was using comes from [here](https://jax.readthedocs.io/en/latest/notebooks/How_JAX_primitives_work.html#transposition), if there's a better way to state this please feel free to suggest something!

File: python/dex/interop/jax.py (hunk 2 above)
Comment (DHHZ) [reply python/dex/interop/jax.py:345]: Right, I see. It seems to me that in order to avoid this we would need to do the closure trick you talked about in the previous PR, that is to extract all intermediate arrays from the computation of `linearized` in the JVP op in order to re-use them here. I'm not sure how we would extract these intermediate arrays with the current version of the Dex foreign API.

Is there a way we could do this? Or am I missing something and is there another way to avoid the checkpointing behaviour?

Comment (apaszke) [reply python/dex/interop/jax.py:345]: Yes, this improvement would require that API and I don't think that we can emulate it in any way.

File: python/dex/interop/jax.py (hunk 1 above)
Comment (apaszke) [reply python/dex/interop/jax.py:328]: Looks great, thanks!

**Review by apaszke — APPROVED** [review]
Awesome, thanks for adding the comment.
